File: src/main/java/net/bms/genera/capability/FaerieInformationProvider.java
package net.bms.genera.capability;

import net.bms.genera.capability.interfaces.IFaerieInformation;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;
import net.minecraftforge.common.capabilities.ICapabilitySerializable;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * Created by benjaminsutter on 7/3/17.
 */
public class FaerieInformationProvider implements ICapabilitySerializable<NBTTagCompound> {

    @CapabilityInject(IFaerieInformation.class)
    static Capability<IFaerieInformation> FAERIE_INFORMATION_CAPABILITY = null;
    private final IFaerieInformation cap = FAERIE_INFORMATION_CAPABILITY.getDefaultInstance();

    @Override
    public boolean hasCapability(@Nonnull Capability<?> capability, @Nullable EnumFacing facing) {
        return capability == FAERIE_INFORMATION_CAPABILITY;
    }

    @Nullable
    @Override
    public <T> T getCapability(@Nonnull Capability<T> capability, @Nullable EnumFacing facing) {
        return capability == FAERIE_INFORMATION_CAPABILITY ? FAERIE_INFORMATION_CAPABILITY.cast(cap) : null;
    }

    @Override
    public NBTTagCompound serializeNBT() {
        return cap.serializeNBT();
    }

    @Override
    public void deserializeNBT(NBTTagCompound nbt) {
        cap.deserializeNBT(nbt);
    }
}

File: src/main/kotlin/net/bms/genera/capability/FaerieInformationProvider.kt
package net.bms.genera.capability

import net.bms.genera.capability.interfaces.IFaerieInformation
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.CapabilityInject
import net.minecraftforge.common.capabilities.ICapabilitySerializable

/**
 * Created by benjaminsutter on 7/3/17.
 */
class FaerieInformationProvider : ICapabilitySerializable<NBTTagCompound> {
    private val cap = FAERIE_INFORMATION_CAPABILITY!!.defaultInstance

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean {
        return capability === FAERIE_INFORMATION_CAPABILITY
    }

    override fun <T> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
        return if (capability === FAERIE_INFORMATION_CAPABILITY) FAERIE_INFORMATION_CAPABILITY!!.cast<T>(cap) else null
    }

    override fun serializeNBT(): NBTTagCompound {
        return cap!!.serializeNBT()
    }

    override fun deserializeNBT(nbt: NBTTagCompound) {
        cap!!.deserializeNBT(nbt)
    }

    companion object {
        @CapabilityInject(IFaerieInformation::class)
        internal var FAERIE_INFORMATION_CAPABILITY: Capability<IFaerieInformation>? = null
    }
}
